### DISCUSSION TITLE
Record, Model conversion

### DISCUSSION THREAD

**Discussion by navopw** [pocketbase/pocketbase#667] [Q&A]
Hello pocketbase community,

I use pocketbase as a framework to build a media storage platform.

What's the best practice for using structs (model) instead of records?

I really would like to access my media as struct and not with the GetValue methods from record.

```go
func MediaGetHandler(c echo.Context) error {
	user, _ := c.Get(apis.ContextUserKey).(*models.User)
	id := c.QueryParam("id")

	// Validation
	if id == "" {
		return c.JSON(http.StatusBadRequest, util.ErrorResponse{
			Message: "missing id",
		})
	}

	// Find media
	col := service.GetMediaCollection(service.App)
	rec, err := service.App.Dao().FindFirstRecordByData(col, "id", id)

	// User check
	if rec.GetStringDataValue("user_id") != user.Id {
		return c.JSON(http.StatusUnauthorized, util.ErrorResponse{
			Message: "unauthorized",
		})
	}

	// Serve
	mime := rec.GetStringDataValue("mime")
	err = service.FileSystem.Serve(c.Response(), id, "media."+util.GetExtensionFromMimeType(mime))
	c.Response().Header().Set("Content-Type", mime)
	if err != nil {
		return c.JSON(http.StatusBadRequest, util.ErrorResponse{
			Message: "failed to serve media",
			Error:   err.Error(),
		})
	}

	return nil
}
```

**Comment by ganigeorgiev** [comment]
You can create a wrapper struct around `models.Record` that has for example methods like `fromRecord` and `toRecord`

OR

create your own typed `Media` struct that implements the `models.Model` interface, eg. something like:

```go
var (
	_ models.Model = (*Media)(nil)
)

type Media struct {
	models.BaseModel

	UserId string `db:"user_id" json:"user_id"`
	Mime   string `db:"mime" json:"mime"`
}

func (m *Media) TableName() string {
	return "media" // the name of your collection
}
```

The base `daos.Dao` operate on the `models.Model` interface so you'll be able to use methods like `dao.Save(model)` to create/update your media record in the database.

For fetch/query, you could use `dao.ModelQuery(&Media{})` to instantiate a `dbx.SelectQuery` instance. For simple examples how to create query helpers you could check the [daos](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.7.6/daos) sub package (`daos/admin.go` is fairly simple to follow and you can use it as a blueprint - https://github.com/pocketbase/pocketbase/blob/master/daos/admin.go)

**Reply by navopw** [reply]
Thank you so much!

PocketBase is an amazing project and I really like to build services with the framework.

**Comment by dschissler** [comment]
@ganigeorgiev

> 	_ models.Model = (*Media)(nil)

I really have no idea what is going on with this wizardry.  I think that not knowing how to read it is stopping me from being able to figure it out.

**Reply by ganigeorgiev** [reply]
This adds extra compile time check to ensure that the `Media` struct is compatible with the `models.Model` interface. The underscore is a special "blank" variable that practically will be discarded during compilation (aka. the value shoudn't take any space).

**Reply by dschissler** [reply]
Thanks.  So is there any reason to use the collection methods that return records?  Is it for JSON marshaling for API?  It just looks like something that I'd never want to use a single time in my routes because it's just worst than having a type.

Another question is that for only selecting records (no writes), is there any gotchas that could come up with passing the raw DB object into another query builder?  It seems that most systems use the `db` tag for the field name, so I'm thinking that is for interoperability.  It would be nice if there was that option.  I'm wondering if perhaps some query hooks get missed or something like that.

**Reply by ganigeorgiev** [reply]
> Thanks. So is there any reason to use the collection methods that return records? Is it for JSON marshaling for API? It just looks like something that I'd never want to use a single time in my routes because it's just worst than having a type.

It depends. Personally I don't mind working with the dynamic structure of the model. The main benefit is that It is easier to maintain and you don't have to keep it in sync with your collection changes. If you frequently operate with the same collection over and over than Yes, probably it would be better and more readable to have a fixed typed struct. But that's up to the developers and their use case.

---

> Another question is that for only selecting records (no writes), is there any gotchas that could come up with passing the raw DB object into another query builder? It seems that most systems use the db tag for the field name, so I'm thinking that is for interoperability. It would be nice if there was that option. I'm wondering if perhaps some query hooks get missed or something like that.

I'm not sure that I understand the question. For db builder we use a forked version of ozzo-dbx. Some examples could be found in the readme of the package https://github.com/pocketbase/dbx#getting-started and its [godoc](https://pkg.go.dev/github.com/pocketbase/dbx). The PocketBase `OnModel*` event hooks fires only if you perform the query via the `Dao` abstraction. You could see how we attach the hooks in https://github.com/pocketbase/pocketbase/blob/9cbb2e750e61884670f093679474ade47d429caf/core/base.go#L748-L776

**Reply by dschissler** [reply]
@ganigeorgiev I'm having trouble bridging the gap between custom models and normal records when it comes to hook events.  It seems that it's either one or the other.  So If I'm using my own models then these hooks are never run when modifying from the API.  So it seems that with custom models that it's only safe to use it when selecting and not for hooks.

**Reply by ganigeorgiev** [reply]
@dschissler The hooks works only with the internal models. You have to "cast" the hook model to your custom struct if you want to use it, eg. with a factory method `customModel.fromRecord(record)`.

**Reply by dschissler** [reply]
Ok, so if I understand this correctly then I wasted a lot of time with that.

**Reply by ganigeorgiev** [reply]
@dschissler Both PocketBase and Go comes with some limitations. You can create your own event hook wrappers that returns your specific custom model, but there is no easy way even with generics to "cast" automatically the Record to your custom struct. Also please note that the event hooks are part of an "execution chain", which means that you can have more than one hook listener, where each listener is fired one by one, allowing you to modify/adjust the record before continuing with the execution (and eventually stopping if `hook.StopPropagation` is returned).

**Reply by dschissler** [reply]
I think that the strategies and limitations should be spelled out better.  Maybe this is something typical of generic API endpoints that I should have known, but never used.  Maybe I should have known more about Golang.

I'm thinking now that I'll do everything generically with records, and then if I want to attach some helper methods or temporary private fields then I'll convert it into a record parent struct.  I think though that I'll leave that for convenience later on, and only for reading.

With the rough edges Pocketbase will have saved me lots of time by the end of the project, but much less than it could have if it was a stable project.  I think that newcomers can save a lot of time if they're nudged in the correct direction as well as dissuaded from going down the path that I went.

**Reply by dschissler** [reply]
@ganigeorgiev I'm thinking that my life would have been better if I had never never seen the [custom models](https://pocketbase.io/docs/custom-models/).  All that it did was to spin me off on a gigantic time waste.

What do you think of something like a notice that this is not recommended for most cases, or that it's experimental.  It definitely needs a lot more thought to flesh it out.  It needs some kind of magic helpers, or something.

In particular I think that it should be very easy to find a notice that if a record is changed that this will come into the model events as a record, and if a custom model was changed then it appears as a custom model.  So that there is a big oopsy doopsy design wall if you're planning on doing important and necessary hooks in both ways.

That is not a kindly brick wall to suddenly hit.

So inevitably I think that it could be fine though because there are so many ways to use Pocketbase, and people can simply disable entire modes to use only a portion of the system.  Though, that needs to be communicated better about what they're giving up and what they're getting.

**Reply by ganigeorgiev** [reply]
@dschissler I don't understand what you propose. Suggestions for improving the documentations are welcomed. The site docs were recently published in https://github.com/pocketbase/site.

The custom model structs are just an option. It is up to the developers whether they want to use them or not. Personally I don't need them, but some users prefer to query their data this way (this is also stated at the beginning in the docs - _"The available [models.Record and its helpers](https://pocketbase.io/docs/record-methods) should be enough for most cases"_).

The hooks works only with the internal models. There is no easy solution when you are dealing with dynamic runtime defined schema and a strongly typed compiled language. Even the Dart SDK has similar "issue", for example check - https://github.com/pocketbase/dart-sdk#accessing-recordmodel-dynamic-fields. The [JS SDK allows returning custom types](https://github.com/pocketbase/js-sdk/tree/master#custom-record-types), but this is because of a "hack"-y workaround possible in TypeScript.

**Reply by dschissler** [reply]
Due to these dynamic runtime limitations in Golang I think that it should be spelled out better what the developer is getting themselves into should they go that route.  Having just wasted a lot of time with that I'm not in a good position to improve the docs right now.  Also although my path seems simple and clear now, I'm still not sure what normal difficulties that I'll run into.

For example, I'm expecting to run into there being no way to abstractly call record hooks.  So there will be an ugly switch buried away that calls specific hook functions based upon the collection name string.  It's a bit ugly, unpure and brittle, but it shouldn't actually affect anything much.

So I still don't 100% know what to do, but I think that I'll be fine now that I'm no longer going down the dark path.
